tools/pbprint.cc

```diff
@@ -3,110 +3,131 @@
-                    return;
-                } else {
-                    if (frame_buffer_index < size) {
-                        /* We haven't received enough data to decode this message */
-                        return;
-                    }
-                }
-                if (msg.MergeFromCodedStream(&input)) {
-                    msg.PrintDebugString();
-                    msg.Clear();
-
-                    frame_buffer_index = 0;
-                    std::cout << std::endl;
-
-                    if (!input.ConsumedEntireMessage()) {
-                        std::cerr << "Entire message not consumed. Number of extra bytes: " <<
-                            input.BytesUntilLimit() << std::endl;
-                    }
-                } else {
-                    std::cerr << "Decode of protobuf message failed." << std::endl;
-                    std::cerr << "Current buffer size: " << frame_buffer_index << std::endl;
-                    std::cerr << "Size after unstuffing: " << unstuff_size << std::endl;
-                    std::cerr << "Decoded varint prefix: " << size << std::endl;
-                    std::cerr << "message buffer contents " << std::endl;
-                    std::cerr << std::hex;
-                    for (size_t i = 0; i < unstuff_size; ++i) {
-                        std::cerr << std::setfill('0') << std::setw(2) << std::right <<
-                            static_cast<int>(sim_buffer[i] & 0xff) <<  " ";
-                    }
-                    std::cerr << std::dec << std::endl;
-
-                    if (unstuff_size > size) {
-                        /* The size of the data stored in the buffer exceeds the
-                         * size of the message obtained by decoding the first
-                         * varint. We may have started reading in the middle of
-                         * a message so throw away the data. */
-                        frame_buffer_index = 0;
-                    }
-                }
-            };
+            exit(-1);
+        }
 
-            while (scan_index < bytes_transferred) {
-                if (rx_buffer[scan_index++] == packet::frame::PACKET_DELIMITER) {
-                    copy_to_frame_buffer();
-                    decode_message();
+        // Update the location we will write to next time. This is
+        // simultaneously the end of the readable part of serial_buffer.
+        serial_buffer_write += serial_buffer_write_count;
+
+        // Decode as many packets as possible starting from serial_buffer_start.
+        const uint8_t * serial_buffer = serial_buffer_start;
+        while (serial_buffer < serial_buffer_write) {
+            // Compute the remaining length.
+            const size_t serial_buffer_len = serial_buffer_write - serial_buffer;
+
+            // Create an output buffer and decode into it.
+            constexpr size_t PACKET_BUFFER_CAPACITY = 2000;
```

Comment:
Why not define this near the top of the file where you also defined `SERIAL_BUFFER_CAPACITY`?